File: app/MMarket_old/c/src/server/fleas/ftests/references.c
// Copyright 17-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "server/fleas/ftests/references.h"
#include "dmc/cgi.h"
#include "data/flea/fmodels.h"
#include "data/flea/Fmodel.h"
#include "data/Nick.h"
#include "db/quotes.h"
#include "db/nicks.h"
#include "DEFS.h"

char *references_process (AsyncActor *ac, Map *mrq) {
  CGI_GET_STR(rq, mrq);

  if (str_eq(rq, "nickList")) {
    char *getName (Nick *nick) { return nick_name(nick); }
    // Map[Js]
    Map *rp = map_new();
    map_put(rp, "nickList", arr_to_js(
      arr_map(nicks_selected_list(), (FCOPY)getName),
      (FTO)js_ws
    ));
    return cgi_ok(rp);
  }
  if (str_eq(rq, "chartData")) {
    CGI_GET_STR(modelId, mrq)
    CGI_GET_STR(nickName, mrq)
    CGI_GET(Darr *, params, darr_from_js, mrq)
    // Map[Js]
    Map *rp = map_new();

    char * error = "";
    void fn () {
      Qtable *opens = opt_nget(quotes_opens());
      if (!opens) { error = "Quotes data base can not be read"; return; }
      double *nick_opens = opt_nget(qtable_nick_values(opens, nickName));
      if (!nick_opens) { error = "Nick not found"; return; }

      Qtable *closes = opt_nget(quotes_closes());
      if (!closes) { error = "Quotes data base can not be read"; return; }
      double *nick_closes = opt_nget(qtable_nick_values(closes, nickName));
      if (!nick_closes) { error = "Nick not found"; return; }

      int fn2 (Fmodel *m) { return str_eq(fmodel_id(m), modelId); }
      Fmodel *model = opt_nget(
        it_find(arr_to_it(fmodels_list()), (FPRED)fn2)
      );
      if (!model) { error = "Model not found"; return; }

      map_put(rp, "dates", arr_to_js(quotes_dates(), (FTO)js_ws));
      map_put(rp, "opens", darr_to_js(
        darr_new_c(HISTORIC_QUOTES, nick_opens)
      ));
      map_put(rp, "closes", darr_to_js(
        darr_new_c(HISTORIC_QUOTES, nick_closes)
      ));
      map_put(rp, "refs", darr_to_js(
        darr_new_c(HISTORIC_QUOTES, fmodel_refs(model, nick_closes, params))
      ));
      map_put(rp, "profits", js_wd(
        fmodel_profits(model, nick_opens, nick_closes, params)
      ));
    }
    asyncActor_wait(ac, fn);

    if (*error)
      return cgi_error(error);
    return cgi_ok(rp);
  }

  EXC_ILLEGAL_ARGUMENT("rq", "chartData", rq)
  return NULL; // Unreachable
}

File: app/KtMarket/go/server/pgs/models/tests/references/references.go
// Copyright 14-Jun-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// tests/References page
package references

import (
	"github.com/dedeme/KtMarket/data/model"
	"github.com/dedeme/KtMarket/data/nick"
	"github.com/dedeme/KtMarket/data/quote"
	"github.com/dedeme/KtMarket/data/strategy"
	"github.com/dedeme/KtMarket/db"
	"github.com/dedeme/ktlib/arr"
	"github.com/dedeme/ktlib/cgi"
	"github.com/dedeme/ktlib/js"
	"github.com/dedeme/ktlib/log"
	"github.com/dedeme/ktlib/thread"
)

func mkArray(n float64) []float64 {
	return []float64{n}
}

func Process(ck string, mrq cgi.T) string {
	rq := js.Rs(mrq["rq"])
	switch rq {
	case "nickList":
		var lst string
		thread.Sync(func() {
			lst = js.Wa(arr.Map(arr.Filter(
				db.NicksTb().Read().List,
				func(n *nick.T) bool {
					return n.IsSel
				},
			),
				func(n *nick.T) string {
					return js.Ws(n.Name)
				},
			))
		})
		return cgi.Rp(ck, cgi.T{
			"nickList": lst,
		})
	case "chartData":
		modelId := js.Rs(mrq["modelId"])
		nickName := js.Rs(mrq["nickName"])
		params := arr.Map(js.Ra(mrq["params"]), js.Rd)

		var ok bool
		var dates []string
		var closes []float64
		var opens []float64
		var refs []float64
		var profits float64

		thread.Sync(func() {
			md, ok2 := arr.Find(model.List(), func(m *model.T) bool {
				return m.Id == modelId
			})
			if !ok2 {
				log.Error("Model '" + modelId + "' not found")
				return
			}

			qs, err := db.QsRead(nickName)
			if err != "" {
				log.Error(err)
				return
			}

			dates = quote.Dates(qs)
			closes = quote.Closes(qs)
			opens = quote.Opens(qs)
			maxs := quote.Maxs(qs)
			st := strategy.New(md, params)
			refs = strategy.Refs(st, closes, -1)

			_, _, _, _, _, _, _, profitss :=
				strategy.Simulation(st, dates, []string{nickName},
					arr.Map(opens, mkArray),
					arr.Map(closes, mkArray),
					arr.Map(maxs, mkArray),
				)
			profits = profitss[0]

			ok = true
		})

		return cgi.Rp(ck, cgi.T{
			"ok": js.Wb(ok),
			"dates": js.Wa(arr.Map(dates, func(e string) string {
				return js.Ws(e)
			})),
			"closes": js.Wa(arr.Map(closes, func(e float64) string {
				return js.Wd(e)
			})),
			"opens": js.Wa(arr.Map(opens, func(e float64) string {
				return js.Wd(e)
			})),
			"refs": js.Wa(arr.Map(refs, func(e float64) string {
				return js.Wd(e)
			})),
			"profits": js.Wd(profits),
		})

	default:
		panic("Value of rq (" + rq + ") is not valid")
	}
}
